From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] reduce PalettedContainer lookup in collisions

Avoid multiple PalettedContainer#get call in CollisionUtil hotpath

License: GPL-3.0-only (https://www.gnu.org/licenses/gpl-3.0.html)

Additional term under GNU GPLv3 section 7(b):

Any copy or modified version containing a substantial portion of the
original contributions in this patch must preserve the following
author attribution in the patch header, source code, or another
prominent legal notice accompanying the corresponding source:

Original patch by HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>

diff --git a/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java b/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
index 54b2492c3ab6bf9ff52cec53dc4635f725f9eb4f..e100d13ced1b6635188320ad908d6625c83c5d2f 100644
--- a/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
+++ b/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
@@ -1980,6 +1980,14 @@ public final class CollisionUtil {
                     final int minYIterate = currChunkY == minChunkY ? (minBlockY & 15) + sectionAdjust : 0;
                     final int maxYIterate = currChunkY == maxChunkY ? (maxBlockY & 15) - sectionAdjust : 15;
 
+                    // Leaf start - reduce PalettedContainer lookup in collisions
+                    if (minXIterate > maxXIterate || minYIterate > maxYIterate || minZIterate > maxZIterate) {
+                        continue;
+                    }
+
+                    final PalettedContainer.Data<BlockState> blocksData = blocks.leaf$getDataAcquire();
+                    // Leaf end - reduce PalettedContainer lookup in collisions
+
                     for (int currY = minYIterate; currY <= maxYIterate; ++currY) {
                         final int blockY = currY | (currChunkY << 4);
                         for (int currZ = minZIterate; currZ <= maxZIterate; ++currZ) {
@@ -1995,7 +2003,7 @@ public final class CollisionUtil {
                                     continue;
                                 }
 
-                                final BlockState blockData = blocks.get(localBlockIndex);
+                                final BlockState blockData = blocks.leaf$getFromData(blocksData, localBlockIndex); // Leaf - reduce PalettedContainer lookup in collisions
 
                                 if (((CollisionBlockState)blockData).moonrise$emptyContextCollisionShape()) {
                                     continue;
